lib/twirp/encoder.ex

```diff
@@ -32,19 +32,19 @@ defmodule Twirp.Encoder do
         {:error, e}
     end
   end
+
   def decode(map, input, @json <> _) do
-    map_with_atoms =
-      map
-      |> Enum.map(fn {key, v} ->
-        k = if is_binary(key), do: String.to_existing_atom(key), else: key
-        {k, v}
-      end)
-      |> Enum.into(%{})
-
-    {:ok, input.new(map_with_atoms)}
-  rescue
-    e ->
-      {:error, e}
+    case Jason.encode(map) do
+      {:ok, bytes} ->
+        decode(bytes, input, @json)
```

Comment:
I'm basically encoding the map back to string and calling the function above this one, which will decode the string back to map. I could also update the existing logic to decode to consider arbitrary depths, but I bet Jason does this faster than I would.

I'm not sure if my understanding is correct, but if this map was previously decoded from a string, we might not need to decode and save some time decoding only here directly with Jason?

Reply:
It's going to be faster and more memory efficient to "decode" this recursively. The only time we hit this is if we're in a phoenix app. That's because Phoenix uses the parser / decoder plug thing and has already read the body and converted it into a map with string keys. In order to get this to work with the protobuf structs we need to map the string keys into atom keys.

Reply:
At first I started with a solution just like you did in the other PR, but I wasn't sure it would be faster. Would have just opened an issue, but this code was already so simple to arrange I figured I would open a PR directly and ask.

Thanks for the explanation, and looking so promptly into this.